api/keystore/service.go

```diff
@@ -301,6 +301,52 @@ func (ks *Keystore) ImportUser(r *http.Request, args *ImportUserArgs, reply *Imp
 	return nil
 }
 
+// DeleteUserArgs are arguments for passing into DeleteUser requests
+type DeleteUserArgs struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+// DeleteUserReply is the response from calling DeleteUser
+type DeleteUserReply struct {
+	Success bool `json:"success"`
+}
+
+// DeleteUser deletes user with the provided username and password.
+func (ks *Keystore) DeleteUser(_ *http.Request, args *DeleteUserArgs, reply *DeleteUserReply) error {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+
+	ks.log.Verbo("DeleteUser called with %s", args.Username)
+
+	if args.Username == "" {
+		return errEmptyUsername
+	}
+
+	// check if user exists and valid user.
+	usr, err := ks.getUser(args.Username)
+	switch {
+	case err != nil || usr == nil:
+		return fmt.Errorf("user doesn't exists: %s", args.Username)
+	case !usr.CheckPassword(args.Password):
+		return fmt.Errorf("incorrect password for user %q", args.Username)
+	}
+
+	// TODO: user is deleted from user db but returned an error due to some other problem.
+	// valid scenario ? discuss with team.
+
+	// delete from user db.
+	if err := ks.userDB.Delete([]byte(args.Username)); err != nil {
+		return err
+	}
```

Comment:
We should also delete the user's data from the database. If we are deleting multiple keys, which we would now be, we should place all these deletions into a batch and then execute that.

This function will essentially be the same as Export and Import combined. Where the Puts in Import are Deletes.

Reply:
understood. Does batch used to achieve Atomicity ? 

couple of questions on batch operation in import api, 
1. does `userBatch.Put` actually executes or just record the action to be performed ? 
2.  returned error is not checked for `dataBatch.Put`, any specific reason ?

   ```
  userBatch := ks.userDB.NewBatch()
	if err := userBatch.Put([]byte(args.Username), usrBytes); err != nil {
		return err
	}

	userDataDB := prefixdb.New([]byte(args.Username), ks.bcDB)
	dataBatch := userDataDB.NewBatch()
	for _, kvp := range userData.Data {
		dataBatch.Put(kvp.Key, kvp.Value)
	}

	if err := atomic.WriteAll(dataBatch, userBatch); err != nil {
		return err
	}

```

Reply:
Used batch logic.